lib/bots/keyframebot.ts

```diff
@@ -0,0 +1,684 @@
+import { AlertLevel, LogLevel } from '../utils/logger';
+import TypedError = require('typed-error');
+import resinSdk = require('resin-sdk');
+
+const exec: (command: string, options?: any) => Promise<{}> = Promise.promisify(ChildProcess.exec);
+const tempMkdir = Promise.promisify(track().mkdir);
+const tempCleanup = Promise.promisify(cleanup);
+
+/**
+ * A custom HTTP error to pass back in event of deployment failure.
+ */
+class HTTPError extends TypedError {
+    /** Message error from the Github API. */
+    public httpCode: number;
+    /** Type of the error, to distinguish it from GithubError. */
+    public type = 'HttpError';
+
+    /**
+     * Constructs a new HTTP error.
+     * @param code      HTTP code to set.
+     * @param message   Message to send back as reason error occurred.
+     */
+    constructor(code: number, message: string) {
+        super();
+
+        // Attempt to parse from JSON.
+        this.httpCode = code;
+        this.message = message;
+    }
+}
+
+/**
+ * Details a keyframe for deployment,
+ */
+interface KeyframeDetails  {
+    /** Version of the keyframe. */
+    version: string;
+    /** The environment in which to deploy the keyframe. */
+    environment: string;
+}
+
+/** Port on which the Github ServiceListener will listen. */
+const GithubPort = 7788;
+/** Endpoint for deployment service. */
+const DeployKeyframePath = '/deploykeyframe';
+/** Port for deployment service. */
+const DeployKeyframePort = 7789;
+/** Standard name of a keyframe. */
+const KeyframeFilename = 'keyframe.yml';
+
+/** Environments interface matching names to repositories. */
+interface Environments {
+    [index: string]: string;
+}
+
+/**
+ * Deployment interface for passing to helper methods.
+ */
+interface DeploymentDetails {
+    keyframe: GithubApiTypes.Content;
+    username: string;
+    version: string;
+    owner: string;
+    repo: string;
+    environment: string;
+}
+
+/**
+ * Constructor for the KeyframeBot.
+ */
+export interface KeyframeBotConstructor {
+    /** Github App id. */
+    integrationId: string;
+    /** PEM for requesting data from Github ServiceEmitter. */
+    pem: string;
+    /** Secret webhook for receiving data from Github ServiceListener. */
+    webhookSecret: string;
+    /** The repo where product keyframes are stored. */
+    productRepo: string;
+    /** The valid environments. JSON of form { "environment": "repo", ... } */
+    environments: string;
+}
+
+// KeyframeBot listens for PullRequest and PullRequestReview events on both product
+// and environment repositories. The general flow is:
+// 1. User creates a PR to update the product keyframe.
+// 2. KeyframeBot lints the PR to ensure the keyframe is valid. It sets a status
+//    on the PR (pass/fail), along with an error if one exists.
+// 3. Usual PR flow is carried out, and eventually keyframe is merged to `master`.
+// 4. At some future point, a user requests that the keyframe is deployed to an environment.
```

Comment:
I guess in theory steps 2 and 4 here could be performed by entirely separate bots, and they're only lumped together here for convenience?